src/msw/textctrl.cpp

```diff
@@ -2862,6 +2883,67 @@ bool wxTextCtrl::MSWSetCharFormat(const wxTextAttr& style, long start, long end)
         }
     }
 
+    if ( style.HasFontUnderlined() )
+    {
+        cf.dwMask |= CFM_UNDERLINETYPE;
+        switch ( style.GetUnderlineType() )
+        {
+            case wxTEXT_ATTR_UNDERLINE_NONE:
+                cf.bUnderlineType = CFU_UNDERLINENONE;
+                break;
+            case wxTEXT_ATTR_UNDERLINE_SOLID:
+                cf.bUnderlineType = CFU_UNDERLINE;
+                break;
+            case wxTEXT_ATTR_UNDERLINE_DOUBLE:
+                cf.bUnderlineType = CFU_UNDERLINEDOUBLE;
+                break;
+            case wxTEXT_ATTR_UNDERLINE_WAVE:
+                cf.bUnderlineType = CFU_UNDERLINEWAVE;
+                break;
+        }
+
+#if _RICHEDIT_VER >= 0x0800
+        // The colours are coming from https://docs.microsoft.com/en-us/windows/desktop/api/tom/nf-tom-itextdocument2-geteffectcolor.
+        BYTE colour = 0;
+        wxColour col = style.GetUnderlineColour();
+        if ( col == wxNullColour )
+            colour = 0;
+        else if ( col == wxTheColourDatabase->Find("BLACK") )
```

Comment:
I know this has been already discussed, but why don't we define an array with RGB values and the corresponding MSW values used just as an index to it? It would seem to be much simpler (and definitely shorter) than these chains of `if`s and easier to update, to boot.

Reply:
done